src/main/scala/org/kframework/minikore/interfaces/Interfaces.scala

```diff
@@ -0,0 +1,493 @@
+  }
+
+  object Leaf {
+    def unapply[C](arg: Leaf[C]): Option[C] = Some(arg.contents)
+  }
+
+
+  /**
+    * A Leaf with Product2[CC1, CC2] as its contents. [[pattern.DomainValue]], [[pattern.Variable]] extend this trait.
+    *
+    * @tparam CC1 Type of First Field.
+    * @tparam CC2 Type of Second Field.
+    */
+  sealed trait Leaf2[CC1, CC2] extends Leaf[Product2[CC1, CC2]] with Product2[CC1, CC2] {
+    override def contents: Product2[CC1, CC2] = (_1, _2)
+
+    def build(_1: CC1, _2: CC2): Pattern
+
+    override def build(contents: Product2[CC1, CC2]): Pattern = build(contents._1, contents._2)
+  }
+
+  object Leaf2 {
+    def unapply[CC1, CC2](arg: Leaf2[CC1, CC2]): Option[(CC1, CC2)] = Some(arg._1, arg._2)
+  }
+
+
+  /**
+    * Node with extra member label, respresenting a node's Label. [[pattern.Application]] that extends this trait.
+    *
+    * @tparam L Type of Label.
+    */
+  sealed trait LabeledNode[L] extends Node with Product1[L] {
+    def build(_1: L, args: Seq[Pattern]): Pattern
+
+    override def build(children: Seq[Pattern]): Pattern = build(_1, children)
+  }
+
+  object LabeledNode {
+    def unapply[L](arg: LabeledNode[L]): Option[(L, Seq[Pattern])] = Some(arg._1, arg.args)
+  }
+
+
+  /**
+    * A Node with empty list of Patterns as its args list. [[pattern.Top]], [[pattern.Bottom]] extend this trait.
+    */
+  sealed trait Node0 extends Node {
+    override def args = Seq.empty[Pattern]
+
+    def build(): Pattern
+
+    override def build(children: Seq[Pattern]): Pattern = {
+      assert(children.isEmpty)
+      build()
+    }
+  }
+
+  object Node0 {
+    def unapply(arg: Node0): Boolean = true
+  }
+
+
+  /**
+    * A Node with a single pattern in its args list. Extended by [[pattern.Next]], [[pattern.Not]].
+    */
+  sealed trait Node1 extends Node with Product1[Pattern] {
+    override def args = Seq(_1)
+
+    def build(_1: Pattern): Pattern
+
+    override def build(children: Seq[Pattern]): Pattern = {
+      assert(children.size == 1)
+      build(children.head)
+    }
+  }
+
+  object Node1 {
+    def unapply(arg: Node1): Option[Pattern] = Some(arg._1)
+  }
+
+
+  /**
+    * A Node with two Patterns in its args list. Extended by [[pattern.Or]], [[pattern.And]], [[pattern.Implies]], [[pattern.Equals]], [[pattern.Rewrite]].
+    */
+  sealed trait Node2 extends Node with Product2[Pattern, Pattern] {
+    def build(_1: Pattern, _2: Pattern): Pattern
+
+    override def args = Seq(_1, _2)
+
+    override def build(children: Seq[Pattern]): Pattern = {
+      assert(children.size == 2)
+      build(children.head, children(1))
+    }
+  }
+
+  object Node2 {
+    def unapply(arg: Node2): Option[(Pattern, Pattern)] = Some(arg._1, arg._2)
+  }
+
+
+  /**
+    * Extends [[Node2]], and only allows [[pattern.Variable]] as the first element, and [[pattern.Pattern]] in its args list. Extended by [[pattern.Exists]], [[pattern.ForAll]].
+    */
+  sealed trait BinderNode extends Node2 {
```

Comment:
I guess this is OK for now, although note that we want to allow users to define their own symbols to be binders, and those symbols may have more than two arguments and the binder may act in only some of the arguments.  Consider, for example, the `let x = e in e'` construct.  The first argument is bound in the third only.